File: typescript/src/problems/s144_binary_tree_preorder_traversal.ts
import { TreeNode } from '~/libs/binary_tree';

/**
 * Runtime: 82ms
 * Memory Usage: 44.7 MB
 * @link https://leetcode.com/problems/binary-tree-preorder-traversal/
 */
export function preorderTraversal(root: TreeNode | null): number[] {
  if (root === null) {
    return [];
  }
  const result: number[] = [];
  const traverse = (node: TreeNode | null) => {
    if (node === null || node.val === null) {
      return;
    }
    result.push(node.val);
    traverse(node.left);
    traverse(node.right);
  };
  traverse(root);
  return result;
}

/**
 * Runtime: 77 ms
 * Memory Usage: 44.8 MB
 */
export function preorderTraversalFromStack(root: TreeNode | null): number[] {
  if (root === null) {
    return [];
  }
  const result: number[] = [];
  const stack: TreeNode[] = [root];
  while (stack.length > 0) {
    let node = stack.pop();
    if (node) {
      if (node.val !== null) {
        result.push(node.val);
      }
      if (node.right !== null) {
        stack.push(node.right);
      }
      if (node.left !== null) {
        stack.push(node.left);
      }
    }
  }
  return result;
}


/**
 * Runtime: 107 ms
 * Memory Usage: 44.7 MB
 */
export function preorderTraversalFromStackTwo(root: TreeNode | null): number[] {
  if (root === null) {
    return [];
  }
  const result: number[] = [];
  const stack: TreeNode[] = [];
  let cur: TreeNode | null = root;
  while (cur !== null || stack.length > 0) {
    while (cur !== null) {
      if (cur.val !== null) {
        result.push(cur.val);
      }
      stack.push(cur);
      cur = cur.left;
    }
    cur = stack.pop()?.right ?? null;
  }
  return result;
}

File: rust/src/problems/s144_binary_tree_preorder_traversal.rs
use std::cell::RefCell;
use std::rc::Rc;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

struct Solution;

impl Solution {
    pub fn preorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        if root.is_none() {
            return vec![];
        }
        fn traverse(node: Option<Rc<RefCell<TreeNode>>>, result: &mut Vec<i32>) {
            if node.is_none() {
                return;
            }
            if let Some(n) = node {
                let node = n.borrow();
                result.push(node.val);
                traverse(node.left.clone(), result);
                traverse(node.right.clone(), result);
            }
        }
        let mut result: Vec<i32> = vec![];
        traverse(root, &mut result);
        result
    }

    pub fn preorder_traversal_from_stack(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        if root.is_none() {
            return vec![];
        }
        let mut stack = vec![root];
        let mut result: Vec<i32> = vec![];
        while let Some(item) = stack.pop() {
            if let Some(value) = item {
                let node = value.borrow();
                result.push(node.val);
                stack.push(node.right.clone());
                stack.push(node.left.clone());
            }
        }
        result
    }

    pub fn preorder_traversal_from_stack_two(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        if root.is_none() {
            return vec![];
        }
        let mut stack = vec![];
        let mut result: Vec<i32> = vec![];
        let mut current = root;
        while current.is_some() || !stack.is_empty() {
            while let Some(item) = current {
                let node = item.borrow();
                result.push(node.val);
                stack.push(node.right.clone());
                current = node.left.clone();
            }
            if let Some(item) = stack.pop() {
                current = item;
            }
        }
        result
    }
}

fn create_binary_tree(nums: Vec<Option<i32>>, index: usize) -> Option<Rc<RefCell<TreeNode>>> {
    if index < nums.len() {
        match nums[index] {
            None => return None,
            Some(val) => {
                let node = Rc::new(RefCell::new(TreeNode::new(val)));
                node.borrow_mut().left = create_binary_tree(nums.clone(), 2 * index + 1);
                node.borrow_mut().right = create_binary_tree(nums.clone(), 2 * index + 2);
                return Some(node);
            }
        }
    }
    None
}

struct Example {
    input: Vec<Option<i32>>,
    output: Vec<i32>,
}

#[test]
pub fn test() {
    let examples = vec![
        Example {
            input: vec![
                Some(3),
                Some(9),
                Some(20),
                Some(11),
                Some(13),
                Some(6),
                Some(7),
                None,
                None,
                Some(15),
                Some(7),
            ],
            output: vec![3, 9, 11, 13, 15, 7, 20, 6, 7],
        },
        Example {
            input: vec![Some(1), None, Some(2)],
            output: vec![1, 2],
        },
        Example {
            input: vec![Some(3), Some(9), Some(20), None, None, Some(15), Some(7)],
            output: vec![3, 9, 20, 15, 7],
        },
        Example {
            input: vec![
                Some(1),
                Some(2),
                Some(3),
                Some(4),
                Some(5),
                Some(6),
                Some(6),
                Some(6),
                Some(6),
            ],
            output: vec![1, 2, 4, 6, 6, 5, 3, 6, 6],
        },
    ];
    for example in examples {
        let root = create_binary_tree(example.input, 0);
        let output = Solution::preorder_traversal_from_stack_two(root);
        assert_eq!(output, example.output);
    }
}
